src/main/scala/vectorpipe/osm/internal/ProcessOSM.scala

```diff
@@ -0,0 +1,756 @@
+      // present and use those
+      history
+        .where('type === "relation")
+        .repartition('id)
+        .select(
+          'id,
+          when(!'visible and (lag('tags, 1) over idByUpdated).isNotNull,
+            lag('tags, 1) over idByUpdated)
+          .otherwise('tags) as 'tags,
+          compressMemberTypes('members) as 'members,
+          'changeset,
+          'timestamp,
+          (lead('timestamp, 1) over idByUpdated) as 'validUntil,
+          'uid,
+          'user,
+          'version,
+          'visible)
+    }
+  }
+
+  /**
+    * Construct all geometries for a set of elements.
+    *
+    * @param elements DataFrame containing node, way, and relation elements
+    * @return DataFrame containing geometries.
+    */
+  def constructGeometries(elements: DataFrame): DataFrame = {
+    import elements.sparkSession.implicits._
+
+    val nodes = ProcessOSM.preprocessNodes(elements)
+
+    val nodeGeoms = ProcessOSM.constructPointGeometries(nodes)
+      .withColumn("minorVersion", lit(0))
+
+    val wayGeoms = ProcessOSM.reconstructWayGeometries(elements, nodes)
+
+    val relationGeoms = ProcessOSM.reconstructRelationGeometries(elements, wayGeoms)
+
+    nodeGeoms
+      .union(wayGeoms.where(size('tags) > 0).drop('geometryChanged))
+      .union(relationGeoms)
+  }
+
+  /**
+    * Construct point geometries. "Uninteresting" nodes are not included, so this is not suitable for way/relation
+    * assembly.
+    *
+    * @param nodes DataFrame containing nodes
+    * @return Nodes as Point geometries
+    */
+  def constructPointGeometries(nodes: DataFrame): DataFrame = {
+    import nodes.sparkSession.implicits._
+
+    val ns = preprocessNodes(nodes)
+      .where(size('tags) > 0)
+
+    ns
+      // fetch the last version of a node within a single changeset
+      .select('changeset, 'id, 'version, 'timestamp)
+      .groupBy('changeset, 'id)
+      .agg(max('version).cast(IntegerType) as 'version, max('timestamp) as 'updated)
+      .join(ns.drop('changeset), Seq("id", "version"))
+      .select(
+        lit(NodeType) as '_type,
+        'id,
+        st_makePoint('lon, 'lat) as 'geom,
+        'tags,
+        'changeset,
+        'updated,
+        'validUntil,
+        'visible,
+        'version)
+  }
+
+  /**
+    * Reconstruct way geometries.
+    *
+    * Nodes and ways contain implicit timestamps that will be used to generate minor versions of geometry that they're
+    * associated with (each entry that exists within a changeset).
+    *
+    * @param _ways        DataFrame containing ways to reconstruct.
+    * @param _nodes       DataFrame containing nodes used to construct ways.
+    * @param _nodesToWays Optional lookup table.
+    * @return Way geometries.
+    */
+  def reconstructWayGeometries(_ways: DataFrame, _nodes: DataFrame, _nodesToWays: Option[DataFrame] = None)(implicit cache: Caching = Caching.none, cachePartitions: Option[Int] = None): DataFrame = {
```

Comment:
I initially added the `implicit Caching` options to facilitate debugging interim outputs; it probably makes sense to remove them now, esp. since they're not used by any of the construction methods.

Reply:
(`Caching.scala` can potentially go away entirely.)

Reply:
Ah, okay. I wasn't sure if the `Caching` was doing something more involved. I will remove those then.